File: sandwich.py
def bread(func):
    def wrapper(*args, **kwargs):
        print "</''''''\>"
        func(*args, **kwargs)
        print "<\______/>"
    return wrapper

def layers(number=1, with_cheese=False):
    def wrap(func):
        def wrapped_f(*args, **kwargs):
            kwargs.update({
                'with_cheese': with_cheese
                })
            for _ in range(number):
                func(*args, **kwargs)
        return wrapped_f
    return wrap

def ingredients(func):
    def wrapper(*args, **kwargs):
        print
        with_cheese = kwargs.pop('with_cheese', False)
        if with_cheese:
            print "*cheese*"
        print "#tomatoes#"

        func(*args, **kwargs)
        print "~salad~"
        if with_cheese:
            print "*cheese*"
        print
    return wrapper

@bread
@layers(number=2, with_cheese=True)
@ingredients
def sandwich(food='--ham--'):
    print food

if __name__ == '__main__':
    print
    sandwich()
    print

File: sandwich.go
package main

import (
    "fmt"
)

func bread(fn func(string)) (func(string)){
    return func(s string) {
        fmt.Println("</''''''\\>")
        fn(s)
        fmt.Println("<\\______/>")
    }
}

func layers(number int, fn func(string)) (func(string)) {
    return func(s string) {
        for i := 0; i < number; i++ {
            fn(s)
        }
    }
}

func ingredients(with_cheese bool, fn func(string)) (func(string)) {
    return func(s string) {
        fmt.Println()
        fmt.Println("#tomatoes#")
        fn(s)
        fmt.Println("~salad~")
        if with_cheese {
            fmt.Println("*cheese*")
        }
        fmt.Println()
    }
}

func sandwich(food string) {
    fmt.Println(food)
}

func main() {
    // create filling with ingredients first
    filling := ingredients(true, sandwich)
    // create layers of filling
    layered_filling := layers(2, filling)
    // add the bread
    full_sandwich := bread(layered_filling)
    full_sandwich("--ham--")
}